### DISCUSSION TITLE
[WASM] Issues when trying latest UNO 4.X with WindowsAppSdk + .NET 6 (+ VS2022)

### DISCUSSION THREAD

**Discussion by DierkDroth** [unoplatform/uno#7723] [Q&A]
I encounter these issues when using the official UNO WASM template or after migrating a personal test project to UNO 4.X + WindowsAppSdk + .NET 6

1) WrapPanel
I get this error 
><source file>(466,11,466,20): error CS0104: 'WrapPanel' is an ambiguous reference between 'CommunityToolkit.WinUI.UI.Controls.WrapPanel' and 'Microsoft.UI.Xaml.Controls.WrapPanel'

However, .NET6 + WindowsAppSDK does not have any 'Microsoft.UI.Xaml.Controls.WrapPanel' class (AFAIK). Why am I getting that error?

2) The WASM project holds a reference Dependencies/Frameworks/Microsoft.AspNetCore.App which has a warning sign on it. Not sure why?!? How could I remove that?

3) There is a NuGet package Microsoft.Windows.Compatibility 5.0.0 referenced. This package could not be removed nor updated (there is a 6.0.0 available). How could that reference be removed? If I really needed it: how could it be updated?

4) I get gazillions of warnings like these in the browser console. How could those be resolved?
> The Bindable attribute is missing and the type [Microsoft.UI.Xaml.Controls.CheckBox] is not known by the MetadataProvider. Reflection was used instead of the binding engine and generated static metadata. Add the Bindable attribute to prevent this message and performance issues.

I recall @jeromelaban and I had a conversation about those warnings: they only would show in DEBUG mode. However, since the warnings suggest performance impact: is there a way to properly resolve the underlying issues and/or get rid of warnings even in DEBUG mode?

5) I get the warnings on (4) even on my own classes where I clearly have a 
> [Bindable(BindableSupport.Default)]

attribute on them. However, these classes implement an interface and the instances of those classes are references by the interface they implement ... not sure if that makes a difference.

6) I see tons of these errors in my VS output
> Failed to match method call Microsoft.WebTools.BrowserLink.LivePreview.LivePreviewExtensionFactory.onEditorSelectionChanged

How could those be resolved?

7) I see these sporadic warning in browser (Chrome) console:
> dotnet.js:3265 warning: 2 FS.syncfs operations in flight at once, probably just doing extra work

How could those be resolved?

8) Not sure if related by after seeing (7) I get tons of these errors in browser console
> dotnet.js:2347 Error #1 "TypeError: Cannot read properties of undefined (reading 'DisableFrameReporting')
put_char @ dotnet.js:2347
write @ dotnet.js:2280
write @ dotnet.js:3903
doWritev @ dotnet.js:5462
_fd_write @ dotnet.js:6590
$func18540 @ 00b4b74e:0x23a502
$func7380 @ 00b4b74e:0xfd36d
...

How could those be resolved?

Also: When migrating my legacy project to latest UNO 4.X I no longer get any 'not supported' compile warnings (before I got tons of them). I would be more than happy if this would be factual, however I'm suspicious that these warnings might be turned off. Is there something I should check to make sure these compile time warnings are ON?

**Comment by jeromelaban** [comment]
Thanks Dierk. We should try to avoid issues bundling as it makes it difficult to follow on conversations. I`ll answer in different sections of the discussion.

**Reply by DierkDroth** [reply]
Understand (did not want to spam the place). Thanks

**Comment by jeromelaban** [comment]
About WrapPanel, can you post the actual XAML you are using?

**Reply by DierkDroth** [reply]
Hmm ... sorry for the delay, I thought I posted a reply .. anyway:

I'm not doing any XAML, but am using WrapPanel in code:
`WrapPanel wrapPanel = new WrapPanel()`
... which does not work in UNO WASM due to the ambiguity error above (but does work in plain WindowsAppSKD + .NET 6), so I had to do
`CommunityToolkit.WinUI.UI.Controls.WrapPanel wrapPanel = new CommunityToolkit.WinUI.UI.Controls.WrapPanel()`

**Reply by jeromelaban** [reply]
Got it. Yes the type is present in both at this point and you'll need to specify the namespace explicitly.

**Reply by DierkDroth** [reply]
To clarify: there is no Microsoft.UI.Xaml.Controls.WrapPanel in .NET6 + WindowsAppSDK. So why would it be in UNO?

Or am I wrong?

**Reply by jeromelaban** [reply]
It's present for historical reasons, and cannot be removed.

**Reply by DierkDroth** [reply]
Ok, understand. Thanks

**Comment by jeromelaban** [comment]
> The WASM project holds a reference Dependencies/Frameworks/Microsoft.AspNetCore.App which has a warning sign on it. Not sure why?!? How could I remove that?

This is a Visual Studio-only false error that we cannot troubleshoot at this point. It does not affect the build or execution of projects.

**Reply by DierkDroth** [reply]
Got it. Thanks. Let's hope MS gets their stuff sorted sooner than later...

**Comment by jeromelaban** [comment]
> I get gazillions of warnings like these in the browser console. How could those be resolved?

This is fixed in Uno 4.1 builds, we'll backport the fix to 4.0 soon so that less errors incorrectly show up.

**Reply by DierkDroth** [reply]
Perfect. Any ETA on 4.1?

**Reply by jeromelaban** [reply]
Not at this point, which is why backporting will be useful. The backport will be available soon after https://github.com/unoplatform/uno/pull/7727 is merged.

**Reply by DierkDroth** [reply]
Fair enough. Thanks

**Reply by DierkDroth** [reply]
@jeromelaban I confirm that warnings are gone after upgrading to 4.1.0-dev.169.

However, there is one warning left:
The Bindable attribute is missing and the type [Microsoft.UI.Xaml.Controls.VirtualizingPanelLayout] is not known by the MetadataProvider....

**Reply by DierkDroth** [reply]
@jeromelaban latest Uno.Wasm.Bootstrap 3.2.0-dev.24 still would show this warning:
The Bindable attribute is missing and the type [Microsoft.UI.Xaml.Controls.VirtualizingPanelLayout] is not known by the MetadataProvider....

Is there anything which could be done?

**Reply by jeromelaban** [reply]
This issue is not yet fixed. It is debug only, you can ignore it. Once fixed, it will be by an Uno.UI update, not a bootstrapper update.

**Reply by DierkDroth** [reply]
Thanks for clarification @jeromelaban

**Comment by jeromelaban** [comment]
> I see tons of these errors in my VS output
Failed to match method call Microsoft.WebTools.BrowserLink.LivePreview.LivePreviewExtensionFactory.onEditorSelectionChanged

This is an unknown error at this point, Let's follow up here: https://github.com/unoplatform/uno/issues/7724

**Comment by jeromelaban** [comment]
> I see these sporadic warning in browser (Chrome) console:
> dotnet.js:3265 warning: 2 FS.syncfs operations in flight at once, probably just doing extra work

Let's track this here: https://github.com/unoplatform/uno/issues/7725

**Comment by jeromelaban** [comment]
> Also: When migrating my legacy project to latest UNO 4.X I no longer get any 'not supported' compile warnings (before I got tons of them). I would be more than happy if this would be factual, however I'm suspicious that these warnings might be turned off. Is there something I should check to make sure these compile time warnings are ON?

We actually included more of those in Uno 4.0.x (XAML is now included). Are `UNO0001` warnings disabled in your csproj by any chance ?

**Reply by DierkDroth** [reply]
> Are UNO0001 warnings disabled in your csproj by any chance ?

I strongly suspect so - I'm almost sure. However, I could not find where I would have disabled UNO0001. I definitely did not disable on purpose - may be it 'sneaked' in om the migration process ....

Is there a way to 'pro-actively' turn it on again?

**Reply by jeromelaban** [reply]
There are multiple ways to disable this warning, but generally it's in the csproj with "NoWarn" nodes.

**Reply by DierkDroth** [reply]
Yep, I know. There only was a 'NU...' warning which I removed from the node. Still no success.

Are these compile time warnings dependent on Code Analyzers or Intellicode? I turned those off, since they're getting in may massively and don't provide any real value...

**Reply by jeromelaban** [reply]
They can be disabled that way as well, in a `.editorconfig` file somewhere in the solution.

**Reply by DierkDroth** [reply]
- I don't have any .editorconfig in my WASM solution (it's separate from the WinUI solution at this point)
- turning on analyzers didn't help

Another thought: is this warning dependent on XAML with .NET6/WinUI? I'm not doing any XAML at all ... it's all 'markup code'

**Reply by DierkDroth** [reply]
Ok, I ran more experiments:
- if I replaced the shared project in my custom solution by the shared project from the UNO WinUI solution template + adding something like 
`var o = new TeachingTip();`
to App.xaml.cs then I would get the UNO00001 warnings in code -> good
- however, I don't get those warnings when putting back my custom shared project - which definitely has tons of unsupported features like .TeachingTip -> bad

So, my best guess is that the warnings no longer would work if the size of a (shared) project is beyond a particular point.

**Reply by DierkDroth** [reply]
@jeromelaban I'm still dead in the water since I have no compile time UNO00001 warnings on my code would not know which features in my code are not supported by UNO.

Here is what we have so far:
- it's definitely not my project file nor my solution file, since if I swapped out my **custom shared** project and replaced it by the **template shared** project then I do get compile time warnings in my code and even in the VS editor without changing any thing else
- I do get the UNO00001 compile time warnings in the custom theme styling XAML template though -> so it should work in principle
- however, I simply don't get neither compile time warning nor the VS editor warning in my **custom shared** project

What else could I try?

**Reply by jeromelaban** [reply]
You may try and analyze the contents of a binlog file, to determine if the UNO0001 is being suppressed when the compiler is invoked, but other than this, the warnings are properly working on our side. If you have a more specific repro case, we can take a look.

**Reply by DierkDroth** [reply]
Ok, which exact compiler call should I be looking for? There are lots of tools and compilers called...

**Reply by DierkDroth** [reply]
I got around running some tests using binlog. However, issue persists and I still have no idea why

- I ran positive and negative tests enabling/disabling the UNO0001 warning. The binlog file held UNO0001 in the command lines as expected (meaning: if disabled there was not a trace of UNO0001 being set to ignore on any command line)
- still I only have the warning for the XAML code (which holds the theming) but *not* a single warning in code - although I definitely have tons of code where e.g. TeachingTip is used

Binlog file attached.

What would be next steps to isolate the cause of trouble?

[msbuild.zip](https://github.com/unoplatform/uno/files/7833502/msbuild.zip)
